numpy/polynomial/polynomial.py

```diff
@@ -780,6 +781,93 @@ def polyval(x, c, tensor=True):
     return c0
 
 
+def polyvalfromroots(x, r):
+    """
+    Evaluate a polynomial specified by its roots at points x.
+
+    If `r` is of length `N`, this function returns the value
+
+    .. math:: p(x) = \prod_{n=1}^{N} (x - r_n)
+
+    The parameter `x` is converted to an array only if it is a tuple or a
+    list, otherwise it is treated as a scalar. In either case, either `x`
+    or its elements must support multiplication and addition both with
+    themselves and with the elements of `r`.
+
+    If `r` is a 1-D array, then `p(x)` will have the same shape as `x`.  If
+    `r` is multidimensional, then the shape of the result depends on the
+    value of `tensor`. If `tensor` is true the shape will be r.shape[1:] +
+    x.shape. If `tensor` is false the shape will be r.shape[1:]. Note that
+    scalars have shape (,).
+
+    .. versionadded:: 1.12
+
+    Parameters
+    ----------
+    x : array_like, compatible object
+        If `x` is a list or tuple, it is converted to an ndarray, otherwise
+        it is left unchanged and treated as a scalar. In either case, `x`
+        or its elements must support addition and multiplication with
+        with themselves and with the elements of `r`.
+    r : array_like
+        Array of roots. If `r` is multidimensional the first index is the
+        root index, while the remaining indices enumerate multiple
+        polynomials. For instance, in the two dimensional case the roots
+        may be thought of as stored in the columns of `r`.
+
+    Returns
+    -------
+    values : ndarray, compatible object
+        The shape of the returned array is described above.
+
+    See Also
+    --------
+    polyroots, polyfromroots, polyval
+
+    Examples
+    --------
+    >>> from numpy.polynomial.polynomial import polyvalfromroots
+    >>> polyvalfromroots(1, [1,2,3])
+    0.0
+    >>> a = np.arange(4).reshape(2,2)
+    >>> a
+    array([[0, 1],
+           [2, 3]])
+    >>> polyvalfromroots(a, [-1, 0, 1])
+    array([[ -0.,   0.],
+           [  6.,  24.]])
+    >>> r = np.arange(-2, 2).reshape(2,2) # multidimensional coefficients
+    >>> r # each column of r defines one polynomial
+    array([[-2, -1],
+           [ 0,  1]])
+    >>> b = np.arange(-2, 3)
+    >>> polyvalfromroots(b, r)
+    array([[-0.,  3.],
+           [-1., -0.],
+           [ 0., -1.],
+           [ 3.,  0.],
+           [ 8.,  3.]])
+    """
+    r = np.array(r, ndmin=1, copy=0)
+    if r.size == 0:
+        return np.ones_like(x)
+    if r.dtype.char in '?bBhHiIlLqQpP':
+        # astype fails with NA
+        r = r + 0.0
+    if isinstance(x, (tuple, list)):
```

Comment:
if x isn't  a tuple or list the rest will fail. `polyval` makes an object array in that case.